### PR TITLE
Add support for non-tarball compressed files

### LINKED ISSUE #3067: Crash when trying to open a compressed file
## Exception
* App Name: Amaze File Manager
* Package: com.amaze.filemanager
* Version: 3.6.7
* User Action: UI Error
* Request: Application crash
* OS: Linux Android 11 - 30
* Device: OnePlus8T
* Model: KB2003
* Product: OnePlus8T_EEA
<details><summary><b>Crash log </b></summary><p>

java.lang.NullPointerException: getCompressorInstance(requireContext(), this) must not be null
 at com.amaze.filemanager.ui.fragments.CompressedExplorerFragment.onViewCreated(CompressedExplorerFragment.kt:235)
 at androidx.fragment.app.Fragment.performViewCreated(Fragment.java:2974)
 at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:543)
 at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:282)
 at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2177)
 at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2088)
 at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1990)
 at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
 at android.os.Handler.handleCallback(Handler.java:938)
 at android.os.Handler.dispatchMessage(Handler.java:99)
 at android.os.Looper.loop(Looper.java:233)
 at android.app.ActivityThread.main(ActivityThread.java:8030)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:631)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:978)

</details>
<hr>

Reporter: None

### PR DESCRIPTION
**Pull request by TranceLove**
## Description
Fixes #3067.

This PR adds support for various non-tarball compressed formats, including gzip, bzip, lzma and xz.

Note that these compression format has no way to tell the original file's uncompressed size and modification time without decompressing (as we know of), hence file size will show up as zero in CompressedExplorerFragment, and decompressed file will only use original file's last modification timestamp.

#### Issue tracker   
Fixes #3067

#### Automatic tests
- [x] Added test cases
  
#### Manual tests
- [x] Done  
  
- Device: Pixel 2 emulator
- OS: Android 11

#### Build tasks success  
Successfully running following tasks on local:
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

### CODE REVIEW

File: app/src/main/java/com/amaze/filemanager/filesystem/compressed/CompressedHelper.java [hunk 1]
```diff
@@ -117,15 +132,21 @@ public static Decompressor getCompressorInstance(@NonNull Context context, @NonN
...
-      decompressor = new LzmaDecompressor(context);
+      decompressor = new TarLzmaDecompressor(context);
     } else if (is7zip(type)) {
       decompressor = new SevenZipDecompressor(context);
+    } else if (isXz(type)) {
+      decompressor = new CommonCompressedFileDecompressor(context);
+    } else if (isLzma(type)) {
+      decompressor = new CommonCompressedFileDecompressor(context);
+    } else if (isGzip(type)) {
+      decompressor = new CommonCompressedFileDecompressor(context);
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/filesystem/compressed/CompressedHelper.java:149]: Three branches that do the same thing.

File: app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/compress/GenericCompressedFileHelperTask.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package com.amaze.filemanager.asynchronous.asynctasks.compress
+
+import com.amaze.filemanager.adapters.data.CompressedObjectParcelable
+import com.amaze.filemanager.asynchronous.asynctasks.AsyncTaskResult
+import com.amaze.filemanager.utils.OnAsyncTaskFinished
+import java.util.ArrayList
+
+class GenericCompressedFileHelperTask(
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/compress/GenericCompressedFileHelperTask.kt:28]: This is not "generic" is more like an "unknowable" or "unknown" CompressedHelperTask

**Review by EmmanuelMess — CHANGES_REQUESTED** [review]

**Comment by EmmanuelMess** [comment]
Thanks for updating the tests!

File: app/src/main/java/com/amaze/filemanager/filesystem/compressed/CompressedHelper.java (hunk 1 above)
Comment (EmmanuelMess) [reply app/src/main/java/com/amaze/filemanager/filesystem/compressed/CompressedHelper.java:149]: If this is feasable please add a comment for why all those files use the file decompressor that provides no info about size and modification.

**Comment by EmmanuelMess** [comment]
Trying to open the txt inside this compressed file is broken (compressed was recompressed as zip to allow for upload, please remove from the zip before testing):
[r.txt.zip](https://github.com/TeamAmaze/AmazeFileManager/files/7786428/r.txt.zip)

**Comment by EmmanuelMess** [comment]
Gzip files have broken icon:
image

How come Ubuntu's compressed explorer is able to show Gzipped file's original size:
image

**Comment by TranceLove** [comment]
> How come Ubuntu's compressed explorer is able to show Gzipped file's original size

According to the spec, gzip does stores the file's original size at the [last 4 bytes of the whole stream](https://en.wikipedia.org/wiki/Gzip#cite_ref-4)... unless there is a strong reason to do so, I'd prefer not to walk the whole input stream for the last 4 byte footer. And also, there had been some says about the inaccuracy of this field for gzips > 4GB.

**Comment by EmmanuelMess** [comment]
Very well, if you add that to a comment then it'll be ok to not have the size, etc

**Review by EmmanuelMess — APPROVED** [review]
Works! Tested on emulator running API 29.

**Review by VishalNehra — APPROVED** [review]

**Pull request merged**